drivers/usb_c/tcpc/ucpd_numaker.c

```diff
@@ -0,0 +1,2528 @@
+
+/**
+ * @brief Initializes VBUS threshold and monitor
+ *
+ * @retval 0 on success
+ * @retval -EIO on failure
+ */
+static int numaker_utcpd_vbus_init(const struct device *dev)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	UTCPD_T *utcpd_base = config->utcpd_base;
+	int rc;
+	uint32_t vbvol = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, VBVOL);
+	uint32_t pwrctl = 0;
+
+	/* UTCPD VBUS scale factor */
+	vbvol &= ~UTCPD_VBVOL_VBSCALE_Msk;
+	vbvol |= config->utcpd.vbvol.vbscale.bit;
+	rc = NUMAKER_UTCPD_REG_WRITE_BY_NAME(dev, VBVOL, vbvol);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if (config->eadc.spec_vbus != NULL) {
+		/* Vref */
+		rc = numaker_eadc_vref_init(dev);
+		if (rc < 0) {
+			return rc;
+		}
+
+		/* UTCPD VBUS overvoltage threshold */
+		rc = NUMAKER_UTCPD_VBUS_THRESHOLD_WRITE(
+			dev, VBOVTH, NUMAKER_UTCPD_VBUS_THRESHOLD_OVERVOLTAGE_MV);
+		if (rc < 0) {
+			return rc;
+		}
+
+		/* UTCPD VBUS vSafe5V threshold */
+		rc = NUMAKER_UTCPD_VBUS_THRESHOLD_WRITE(dev, VSAFE5V,
+							NUMAKER_UTCPD_VBUS_THRESHOLD_VSAFE5V_MV);
+		if (rc < 0) {
+			return rc;
+		}
+
+		/* UTCPD VBUS vSafe0V threshold */
+		rc = NUMAKER_UTCPD_VBUS_THRESHOLD_WRITE(dev, VSAFE0V,
+							NUMAKER_UTCPD_VBUS_THRESHOLD_VSAFE0V_MV);
+		if (rc < 0) {
+			return rc;
+		}
+
+		/* UTCPD VBUS stop force discharge threshold */
+		rc = NUMAKER_UTCPD_VBUS_THRESHOLD_WRITE(
+			dev, SPDGTH, NUMAKER_UTCPD_VBUS_THRESHOLD_STOP_FORCE_DISCHARGE_MV);
+		if (rc < 0) {
+			return rc;
+		}
+
+		/* UTCPD VBUS sink disconnect threshold */
+		rc = NUMAKER_UTCPD_VBUS_THRESHOLD_WRITE(
+			dev, SKVBDCTH, NUMAKER_UTCPD_VBUS_THRESHOLD_SINK_DISCONNECT_MV);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	/* Enable UTCPD VBUS voltage monitor so that UTCPD.VBVOL is available */
+	if (config->eadc.spec_vbus != NULL) {
+		pwrctl &= ~UTCPD_PWRCTL_VBMONI_DIS;
+	} else {
+		pwrctl |= UTCPD_PWRCTL_VBMONI_DIS;
+	}
+	/* Disable UTCPD VBUS voltage alarms */
+	pwrctl |= UTCPD_PWRCTL_DSVBAM_DIS;
+	/* Disable UTCPD VBUS auto-discharge on disconnect
+	 * NOTE: UTCPD may not integrate with discharge, so this feature is
+	 *	 disabled and discharge is handled separately.
+	 */
+	pwrctl &= ~UTCPD_PWRCTL_ADGDC;
+	return NUMAKER_UTCPD_REG_WRITE_BY_NAME(dev, PWRCTL, pwrctl);
+}
+
+/**
+ * @brief Initializes UTCPD GPIO pins
+ *
+ * @retval 0 on success
+ * @retval -EIO on failure
+ */
+static int numaker_utcpd_gpios_init(const struct device *dev)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	int rc;
+	const struct gpio_dt_spec *spec;
+
+	/* Configure VBUS detect pin to INPUT to avoid intervening its power measurement */
+	spec = &config->utcpd.gpios.vbus_detect;
+	if (spec->port == NULL) {
+		LOG_ERR("VBUS detect pin not specified");
+		return -ENODEV;
+	}
+	if (!gpio_is_ready_dt(spec)) {
+		LOG_ERR("VBUS detect pin port device not ready");
+		return -ENODEV;
+	}
+	rc = gpio_pin_configure_dt(spec, GPIO_INPUT);
+	if (rc < 0) {
+		LOG_ERR("VBUS detect pin configured to INPUT failed");
```

Comment:
these LOG_ERR in respons to a return code should probably log the code as part of the message, just to give an extra hint if they fail